File: tool/extract_vocabulary.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# List the romanized words a language's lessons use but its dictionary lacks.
#
# Every target-language word a learner can tap needs a gloss, so coverage is
# derived from the lessons rather than maintained by hand.
#
# Usage:
#   ruby tool/extract_vocabulary.rb tamil                 # missing words, one per line
#   ruby tool/extract_vocabulary.rb tamil --all           # every word used
#   ruby tool/extract_vocabulary.rb tamil --json out.json

require_relative 'courses'

module ExtractVocabulary
  module_function

  # Word => number of times it appears, across every lesson of a language.
  def vocabulary(language)
    counts = Hash.new(0)
    Courses.course_files(language).each do |path|
      Courses.read_json(path).fetch('levels', []).each do |level|
        level.fetch('questions', []).each do |question|
          Courses.words_in(question['sentence'].to_s).each { |word| counts[word] += 1 }
          next unless question['type'] == 'multiple_choice'

          question.fetch('options', []).each do |option|
            Courses.words_in(option).each { |word| counts[word] += 1 }
          end
        end
      end
    end
    counts.delete('name') # the {name} placeholder
    counts
  end

  def main(argv)
    language = argv.find { |arg| !arg.start_with?('--') }
    abort 'usage: ruby tool/extract_vocabulary.rb <language> [--all] [--json PATH]' unless language

    counts = vocabulary(language)
    unless argv.include?('--all')
      dictionary_path = Courses::COURSES_DIR / language / 'dictionary.json'
      known = if dictionary_path.exist?
                Courses.read_json(dictionary_path).keys.map { |w| Courses.normalize_word(w) }.to_set
              else
                Set.new
              end
      counts = counts.reject { |word, _| known.include?(word) }
    end

    # Most frequent first: the words worth getting right.
    ordered = counts.sort_by { |word, count| [-count, word] }.to_h

    json_index = argv.index('--json')
    if json_index
      out = argv[json_index + 1]
      Courses.write_json(out, ordered)
      puts "#{ordered.size} words -> #{out}"
    else
      ordered.each_key { |word| puts word }
    end
    0
  end
end

exit ExtractVocabulary.main(ARGV) if $PROGRAM_NAME == __FILE__

File: tool/extract_vocabulary.py
#!/usr/bin/env python3
"""List the romanized words a language's lessons use but its dictionary lacks.

Every target-language word a learner can tap needs a gloss, so coverage is
derived from the lessons rather than maintained by hand.

Usage:
    python3 tool/extract_vocabulary.py tamil              # missing words, one per line
    python3 tool/extract_vocabulary.py tamil --all        # every word used
    python3 tool/extract_vocabulary.py tamil --json out.json
"""
import argparse
import json
import pathlib
import sys
from collections import Counter

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent))
from validate_courses import COURSES_DIR, normalize_word, words_in  # noqa: E402


def vocabulary(language):
    """Word -> number of times it appears, across every lesson of a language."""
    directory = COURSES_DIR / language
    counts = Counter()
    for path in sorted(directory.glob("*.json")):
        if path.name in ("manifest.json", "dictionary.json"):
            continue
        data = json.loads(path.read_text())
        for level in data.get("levels", []):
            for question in level.get("questions", []):
                counts.update(words_in(question.get("sentence", "")))
                if question.get("type") == "multiple_choice":
                    for option in question.get("options", []):
                        counts.update(words_in(option))
    counts.pop("name", None)  # the {name} placeholder
    return counts


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("language")
    parser.add_argument("--all", action="store_true",
                        help="list every word, not just the unglossed ones")
    parser.add_argument("--json", metavar="PATH",
                        help="write {word: occurrences} to PATH instead of stdout")
    args = parser.parse_args()

    counts = vocabulary(args.language)
    if not args.all:
        dictionary_path = COURSES_DIR / args.language / "dictionary.json"
        known = set()
        if dictionary_path.exists():
            known = {normalize_word(w) for w in json.loads(dictionary_path.read_text())}
        counts = Counter({w: n for w, n in counts.items() if w not in known})

    # Most frequent first: the words worth getting right.
    ordered = dict(counts.most_common())
    if args.json:
        pathlib.Path(args.json).write_text(json.dumps(ordered, indent=2) + "\n")
        print(f"{len(ordered)} words -> {args.json}")
    else:
        for word in ordered:
            print(word)


if __name__ == "__main__":
    main()
